src/tactic/rewrite_search/search.lean

```diff
@@ -0,0 +1,208 @@
+of our initial equation.
+`conf` and `rules` determine what rewrites are used to generate new graph vertices.
+Here, the format of a rewrite rule is an expression for rewriting, plus a flag for the
+direction to apply it in.
+`vertices` maps vertex.id to vertex.
+`vmap` maps vertex.pp to vertex.id, so we can quickly find collisions.
+`solving_edge` represents a solution that will prove our target equation.
+It is an edge that would connect the two trees, so `solving_edge.fr` and `solving_edge.to`
+are vertices in different trees.
+-/
+meta structure graph :=
+(conf         : config)
+(rules        : list (expr × bool))
+(vertices     : buffer vertex)
+(vmap         : native.rb_map string ℕ)
+(solving_edge : option edge)
+
+/--
+Construct a graph to search for a proof of a given equation.
+-/
+meta def mk_graph (conf : config) (rules : list (expr × bool)) (eq : expr) : tactic graph :=
+do (lhs, rhs) ← split_equation eq,
+  lhs_pp ← to_string <$> tactic.pp lhs,
+  rhs_pp ← to_string <$> tactic.pp rhs,
+  let lhs_vertex : vertex := ⟨0, lhs, lhs_pp, side.L, none⟩,
+  let rhs_vertex : vertex := ⟨1, rhs, rhs_pp, side.R, none⟩,
+  return ⟨conf, rules, [lhs_vertex, rhs_vertex].to_buffer,
+          native.rb_map.of_list [(lhs_pp, 0), (rhs_pp, 1)], none⟩
+
+variables (g : graph)
+
+namespace graph
+
+/--
+A convenience wrapper around `g.vertices` that fails if the index is out of bounds.
+-/
+meta def get_vertex (i : ℕ) : tactic vertex :=
+if h : i < g.vertices.size then return $ g.vertices.read (fin.mk i h)
+else fail "invalid vertex access"
+
+/--
+Find a list of edges that connect the given edge to the root of its tree.
+The edges are returned in leaf-to-root order, while they are in root-to-leaf direction,
+so if you want them in the logical order you must reverse the returned list.
+-/
+private meta def walk_up_parents : option edge → tactic (list edge)
+| none     := return []
+| (some e) := do
+  v ← g.get_vertex e.fr,
+  edges ← walk_up_parents v.parent,
+  return (e :: edges)
+
+/--
+Returns two lists that represent a solution. The first list is a path from LHS to some
+interior vertex, the second is a path from the RHS to that interior vertex.
+-/
+private meta def solution_paths : tactic (list edge × list edge) :=
+do e ← g.solving_edge,
+  v ← g.get_vertex e.to,
+  path1 ← walk_up_parents g e,
+  path2 ← walk_up_parents g v.parent,
+  match v.side with
+  | side.L := return (path2.reverse, path1.reverse)
+  | side.R := return (path1.reverse, path2.reverse)
+  end
+
+/--
+Add the new vertex and edge to the graph, that can be proved in one step starting
+at a given vertex, with a given rewrite expression.
+For efficiency, it's important that this is the only way the graph is mutated,
+and it only appends to the end of the `vertices` buffer.
+-/
+private meta def add_rewrite (v : vertex) (rw : rewrite) : tactic graph :=
+do pp ← to_string <$> tactic.pp rw.exp,
+  let maybe_id := g.vmap.find pp,
```

Comment:
We seem to match the new expression only by string equality. How do you deal with the case that two expressions have the same `pp` but do not unify, which seems to return an invalid proof here? (There is also the converse: two expressions unify but do not have the same `pp`. Continuing the search would be fine in that case I assume.)

Reply:
Hmm, this is a good point that I didn't consider. I thought about ways to handle this, but I'm not quite sure what to do. You are correct that it's okay if this expression matching misses two expressions that unify, but it's bad if it accepts expressions that actually do not unify. I think the ideal way would be to implement hashing on expressions in such a way that syntactically equivalent expressions had the same hash, and collisions were very unlikely. But, I don't think we can do that efficiently in Lean right now. We could check for unification and carry on, but once two nodes are mistakenly identified with this algorithm, it seems like many nodes will be mistakenly unified. So... the best suggestion I have is just to fail in this case. I at least added a comment to this effect in the overall summary. In practice, I have not run into this in any of the cases I tested at least.

Reply:
Ah, I see my earlier comment comes up here. What's the problem with making `vmap` an `expr_map nat`, i.e. `rb_map expr nat`? Is the problem that you need to match things that are "pp-equal" but not indentical, e.g. they have different but defeq type class instances?

Reply:
Well, I wasn't even aware that it was possible to use an `expr` as the key of an `rb_map`. It is certainly aesthetically superior to use an `expr_map`. So I tried it out, and unfortunately it makes the algorithm about 10x slower on the `rewrite_search.lean` test file, from 2 second compilation to 20 seconds. This map and the buffer are just in the inner loop of this algorithm, and AFAICT doing the expression comparisons is slow. I think to make this work we need to be cranking out all the speed improvements we can get, and in practice `pp` equality seems to be good enough. But ideally this entire map could be a c++ data structure and the inner loop of this logic could be in c++, but I don't want to get that messy in a way that will likely be incompatible with Lean 4. I'd like to revisit this once Lean 4 is working with mathlib though.